src/ImageSharp/Common/Helpers/SimdUtils.HwIntrinsics.cs

```diff
@@ -295,6 +295,190 @@ public static class HwIntrinsics
                 }
             }
+
+            internal static void PackBytesToUInt32SaturateChannel4Reduce(
+                ref ReadOnlySpan<byte> channel0,
+                ref ReadOnlySpan<byte> channel1,
+                ref ReadOnlySpan<byte> channel2,
+                ref Span<byte> dest)
+            {
+                DebugGuard.IsTrue(channel0.Length == dest.Length, nameof(channel0), "Input spans must be of same length!");
+                DebugGuard.IsTrue(channel1.Length == dest.Length, nameof(channel1), "Input spans must be of same length!");
+                DebugGuard.IsTrue(channel2.Length == dest.Length, nameof(channel2), "Input spans must be of same length!");
+
+                if (Avx2.IsSupported)
+                {
+                    int remainder = ImageMaths.ModuloP2(channel1.Length, Vector256<byte>.Count);
+                    int adjustedCount = channel1.Length - remainder;
+
+                    if (adjustedCount > 0)
+                    {
+                        channel0 = channel0.Slice(adjustedCount);
+                        channel1 = channel1.Slice(adjustedCount);
+                        channel2 = channel2.Slice(adjustedCount);
+                        dest = dest.Slice(adjustedCount);
+
+                        PackBytesToUInt32SaturateChannel4(
+                            channel0,
+                            channel1,
+                            channel2,
+                            dest);
+                    }
+                }
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal static void PackBytesToUInt32SaturateChannel4(
+                ReadOnlySpan<byte> channel0,
+                ReadOnlySpan<byte> channel1,
+                ReadOnlySpan<byte> channel2,
+                Span<byte> dest)
+            {
+                int n = dest.Length / Vector256<byte>.Count;
+
+                ref Vector256<byte> source0Base =
+                    ref Unsafe.As<byte, Vector256<byte>>(ref MemoryMarshal.GetReference(channel0));
+                ref Vector256<byte> source1Base =
+                    ref Unsafe.As<byte, Vector256<byte>>(ref MemoryMarshal.GetReference(channel1));
+                ref Vector256<byte> source2Base =
+                    ref Unsafe.As<byte, Vector256<byte>>(ref MemoryMarshal.GetReference(channel2));
+
+                ref Vector256<byte> destBase = ref Unsafe.As<byte, Vector256<byte>>(ref MemoryMarshal.GetReference(dest));
+
+                Vector256<byte> allOnes = Avx2.CompareEqual(Vector256<byte>.Zero, Vector256<byte>.Zero);
+
+                for (int i = 0, j = 0; j < n; i += 1, j += 4)
+                {
+                    Vector256<byte> s0 = Unsafe.Add(ref source0Base, i);
+                    Vector256<byte> s1 = Unsafe.Add(ref source1Base, i);
+                    Vector256<byte> s2 = Unsafe.Add(ref source2Base, i);
+
+                    s0 = Avx2.Permute4x64(s0.AsUInt64(), 0b_11_01_10_00).AsByte();
```

Comment:
If I'm not missing something here we can pack these with less work in a manner similar to how it's done in the jpeg color converter?

https://github.com/SixLabors/ImageSharp/blob/120080b99f3f1be0fc4ecec6906ee683ea999a3d/src/ImageSharp/Formats/Jpeg/Components/Decoder/ColorConverters/JpegColorConverter.FromYCbCrSimdAvx2.cs#L44

Reply:
Yep, it's the same problem.  The unpack operations work in-lane, so whatever starts in the upper lane of the input will be in the upper lanes of the output.  The current code re-permutes after each round to keep things in the right place throughout.  The other options are:

1) Deal with it at the end, by writing the 4 lower lanes to the output first followed by the 4 upper lanes.
2) Rearrange the input so the blocks of 4 bytes (which will become blocks of 16 bytes) will come out paired correctly after the unpacks.

2) will be cheaper since a extract costs the same as a permute, and as with the YCbCr conversion, you can get by with only permuting 3 inputs for 4 outputs.

Reply:
Thought so, thanks for confirming. 👍